File: crates/ferrogate-cloudflare/src/scopes.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-23
// description: Machine-adjacent list of the Cloudflare token permission groups FerroGate needs.

//! Required Cloudflare API-token permission groups (issue #405).
//!
//! This is the machine-adjacent source of truth referenced by
//! [`crate::CloudflareClient::preflight`] when a token authenticates but is
//! under-scoped. The operator-facing prose doc is a sibling sub-issue (#421);
//! this table is what the code names in error messages so the two never drift.
//!
//! An API token is scoped by attaching **permission groups** at the account
//! (or zone) level. FerroGate's Cloudflare integrations collectively require
//! the groups below. A given deployment that only uses a subset of the
//! Cloudflare features can grant the corresponding subset — but the preflight
//! check reports the full foundational set so an operator can provision once.

/// One required token permission group and why FerroGate needs it.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TokenPermissionGroup {
    /// Cloudflare's dashboard name for the permission group.
    pub name: &'static str,
    /// Read / Edit (Cloudflare's Read vs. Edit access level).
    pub access: &'static str,
    /// Which FerroGate Cloudflare sub-system consumes it.
    pub used_by: &'static str,
}

/// The full set of permission groups the FerroGate Cloudflare integrations
/// depend on. Kept in issue-declared order.
pub const REQUIRED_TOKEN_PERMISSION_GROUPS: &[TokenPermissionGroup] = &[
    TokenPermissionGroup {
        name: "AI Gateway",
        access: "Read, Edit",
        used_by: "AI Gateway management + inference proxying",
    },
    TokenPermissionGroup {
        name: "Secrets Store",
        access: "Read, Write",
        used_by: "cf:// secret backend (#417)",
    },
    TokenPermissionGroup {
        name: "D1",
        access: "Read, Edit",
        used_by: "D1-backed state",
    },
    TokenPermissionGroup {
        name: "Workers Scripts",
        access: "Edit",
        used_by: "Worker deployment",
    },
    TokenPermissionGroup {
        name: "Workers R2 Storage",
        access: "Read, Edit",
        used_by: "R2 object storage",
    },
    TokenPermissionGroup {
        name: "API Tokens",
        access: "Write",
        used_by: "minting/revoking bucket-scoped R2 API tokens (#462)",
    },
    TokenPermissionGroup {
        name: "Cloudflare Pages",
        access: "Edit",
        used_by: "Pages deployment",
    },
    TokenPermissionGroup {
        name: "Workflows (Workers Scripts)",
        access: "Write, Edit",
        used_by: "Workflows orchestration",
    },
];

/// The permission-group names, for embedding in a [`crate::CloudflareError`]
/// message. Allocation-free view over [`REQUIRED_TOKEN_PERMISSION_GROUPS`].
pub fn required_group_names() -> Vec<&'static str> {
    REQUIRED_TOKEN_PERMISSION_GROUPS
        .iter()
        .map(|g| g.name)
        .collect()
}

File: packages/cloudflare/src/scopes.ts
/**
 * Required Cloudflare API-token permission groups — slice **S3**.
 *
 * Ported from `crates/ferrogate-cloudflare/src/scopes.rs`. This is the
 * machine-adjacent source of truth `CloudflareClient.preflight()` names when a
 * token authenticates but is under-scoped.
 *
 * An API token is scoped by attaching **permission groups** at the account (or
 * zone) level. A deployment that uses only some Cloudflare features can grant
 * only the corresponding subset — but preflight reports the FULL foundational
 * set on purpose, so an operator provisions once instead of discovering each
 * missing group at its own first use, in production.
 *
 * The strings are Cloudflare dashboard names. A typo here is un-actionable
 * advice and the type system cannot notice it, which is why `test/scopes.test.ts`
 * pins every row verbatim.
 *
 * This module has NO imports, deliberately: it is the leaf of the package's
 * dependency graph (`errors.ts` reads it), so a consumer can adopt just the
 * table.
 */

/** One required token permission group and why FerroGate needs it. */
export interface TokenPermissionGroup {
  /** Cloudflare's dashboard name for the permission group. */
  readonly name: string;
  /** Cloudflare's access level(s) for the group — Read / Edit / Write. */
  readonly access: string;
  /** Which FerroGate Cloudflare subsystem consumes it. */
  readonly usedBy: string;
}

/** The full set of permission groups FerroGate's integrations depend on. */
export const REQUIRED_TOKEN_PERMISSION_GROUPS: readonly TokenPermissionGroup[] = [
  {
    name: "AI Gateway",
    access: "Read, Edit",
    usedBy: "AI Gateway management + inference proxying",
  },
  { name: "Secrets Store", access: "Read, Write", usedBy: "cf:// secret backend" },
  { name: "D1", access: "Read, Edit", usedBy: "D1-backed state, incl. tenant database lifecycle" },
  { name: "Workers Scripts", access: "Edit", usedBy: "Worker deployment" },
  {
    name: "Workers R2 Storage",
    access: "Read, Edit",
    usedBy: "R2 bucket MANAGEMENT (distinct from the S3 key pair)",
  },
  {
    name: "API Tokens",
    access: "Write",
    usedBy: "minting/revoking bucket-scoped R2 API tokens",
  },
  { name: "Cloudflare Pages", access: "Edit", usedBy: "Pages deployment" },
  {
    name: "Workflows (Workers Scripts)",
    access: "Write, Edit",
    usedBy: "Workflows orchestration",
  },
];

/** The permission-group names, in table order, for embedding in an error. */
export function requiredGroupNames(): readonly string[] {
  return REQUIRED_TOKEN_PERMISSION_GROUPS.map((group) => group.name);
}
